evaluate.py

```diff
@@ -78,281 +49,33 @@ def add_batch(self, y_true, y_pred):
         self.y_true.append(y_true)
         self.y_pred.append(y_pred)
 
-    def filter_instances(self, label_idxs, y_true, y_pred):
-        """1. Instances that do not contains labels idxs are removed
-           2. Labels that do not contains in the label idxs are removed
-        """
-        mask_np = np.zeros(self.num_class)
-        mask_np[label_idxs] = 1
-
-        valid_instances_idxs = list()
-        for i, y in enumerate(y_true):
-            if (y * mask_np).sum() > 0:
-                valid_instances_idxs.append(i)
-        return y_true[valid_instances_idxs][:, label_idxs], y_pred[valid_instances_idxs][:, label_idxs]
-
     def eval(self, y_true, y_pred, threshold=0.5):
-        results = []
-
-        for group_idxs, group_name in self.label_groups:
-            result = {'Label Group': group_name, 'Label Size': len(group_idxs)}
-            target_y_true, target_y_pred = self.filter_instances(
-                group_idxs, y_true, y_pred)
-            result['# Instance'] = len(target_y_true)
-
-            # micro/macro f1 of the target groups
-            # micro_f1 = f1_score(y_true=target_y_true, y_pred=target_y_pred > threshold, average='micro')
-            # macro_f1 = f1_score(y_true=target_y_true, y_pred=target_y_pred > threshold, average='macro')
+        result = {
+            'Label Size': self.config.num_classes,
+            '# Instance': len(y_true),
+            'Micro-F1': f1_score(y_true, y_pred > threshold, average='micro'), # sklearn's micro-f1
+            'Macro-F1': macro_f1(y_true, y_pred > threshold) # caml's macro-f1
+        }
+
+        # add metric like P@k, R@k to the result dict
+        pattern = re.compile('(?:P|R)@\d+')
+        for metric in self.config.monitor_metrics:
+            for pr_metric in re.findall(pattern, metric):
+                metric_type, top_k = pr_metric.split('@')
+                top_k = int(top_k)
+                metric_at_k = precision_at_k(y_true, y_pred, k=top_k) if metric_type == 'P' \
+                    else recall_at_k(y_true, y_pred, k=top_k)
+                result[pr_metric] = metric_at_k
```

Comment:
1. Use re.match() instead of re.findall() as we assume a single metric here.
2. Should raise an exception if an invalid metric is specified.

```suggestion
        for metric in self.config.monitor_metrics:
            if re.match('P@\d+', metric):
                top_k = int(metric[2:])
                metric_at_k = precision_at_k(y_true, y_pred, k=top_k)
            elif re.match('R@\d+', metric):
                top_k = int(metric[2:])
                metric_at_k = recall_at_k(y_true, y_pred, k=top_k)
            else:
                raise ValueError(f'Invalid metric: {metric}')
            result[metric] = metric_at_k
```